### PR TITLE
t176

### PR DESCRIPTION
**Pull request by waltdisgrace**

### CODE REVIEW

File: lib/pacemaker/pcmk_output.c [hunk 1]
```diff
@@ -1917,6 +1917,11 @@ pcmk__cluster_status_text(pcmk__output_t *out, va_list args)
         }
     }
 
+    /* Always print verification errors if there are any, even if not requested */
+    if (pcmk_is_set(section_opts, pcmk_section_validate)) {
+        rc = pcmk__verify(scheduler, out, scheduler->input);
+    }
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:1923]: This is a good step to ensure it's working as desired. However for the final output, I don't want to clutter the status display with what may be a long list of problems, so what we can do is create a temporary output object with output type "none", call `pcmk__verify()` regardless of whether the flag is set or not, then use only the return code to display a message in the status. (That will also make it easy to always show it when there's a problem.)

Basically if there's no error, it should say something to the effect that there are no CIB validation issues, and if there is an error, it should say there are CIB validation issues and to run crm_verify for more details.

**Review by kgaillot — COMMENTED** [review]
Good start

**Comment by kgaillot** [comment]
Changed my mind on validate vs verify, updated comment :)

File: lib/pacemaker/pcmk_output.c [hunk 2]
```diff
@@ -1999,6 +2019,21 @@ cluster_status_xml(pcmk__output_t *out, va_list args)
...
+    verify_rc = pcmk__verify(scheduler, verify_out, scheduler->input);
+    pcmk__output_free(verify_out);
+
+    if (verify_rc != pcmk_rc_schema_validation) {
+        if (pcmk_is_set(section_opts, pcmk_section_verify)) {
+            out->info(out, "No verification errors");
+        }
+    } else {
+        out->info(out, "Found verification errors! Run crm_verify for more information.");
+    }
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:2035]: `out->info()` is ignored for XML output. This is where our work on crm_verify will come in handy. For XML, we don't need to create a "none" object, we can pass crm_mon's output object. That should end up with output something like this:

    <summary>
      <stack type="corosync"/>
      <current_dc present="true" version="" name="cluster02" id="2" with_quorum="true" mixed_version="false"/>
      <last_update time=""/>
      <last_change time=""/>
      <nodes_configured number="5"/>
      <verifications code="78" message="Invalid configuration">
        <errors>
          <error>Resource start-up disabled since no STONITH resources have been defined</error>
          <error>Either configure some or disable STONITH with the stonith-enabled option</error>
          <error>NOTE: Clusters with shared data need STONITH to ensure data integrity</error>
          <error>Node pcmk-1 is unclean but cannot be fenced</error>
          <error>Node pcmk-2 is unclean but cannot be fenced</error>
          <error>CIB did not pass schema validation</error>
          <error>Errors found during check: config not valid</error>
        </errors>
      </verifications>
      <resources_configured number="32" disabled="4" blocked="0"/>
      <cluster_options stonith-enabled="true" symmetric-cluster="true" no-quorum-policy="stop" maintenance-mode="false" stop-all-resources="false" stonith-timeout-ms="60000" priority-fencing-delay-ms="0"/>
    </summary>

File: cts/cli/regression.crm_mon.exp [hunk 3]
```diff
@@ -591,6 +591,7 @@ Operations:
 
 Negative Location Constraints:
   * not-on-cluster1	prevents dummy from running on cluster01
+No verification errors
```
Comment (kgaillot) [inline cts/cli/regression.crm_mon.exp:594]: Some of the crm_mon regression tests use `--include=all`, so it's expected that they will start showing the new text. XML output defaults to `--include=all`, so that will show up in even more tests.

File: cts/cli/regression.daemons.exp [hunk 4]
```diff
@@ -27,13 +27,13 @@
   <shortdesc lang="en">Pacemaker controller options</shortdesc>
   <parameters>
     <parameter name="dc-version">
-      <longdesc lang="en">Includes a hash which identifies the exact revision the code was built from. Used for diagnostic purposes.</longdesc>
+      <longdesc lang="en">Includes a hash which identifies the exact changeset the code was built from. Used for diagnostic purposes.</longdesc>
```
Comment (kgaillot) [inline cts/cli/regression.daemons.exp:30]: This was updated in #3325. I'm not sure how you ended up with just some of the changes from that; maybe try exporting your commits as patches and reapply them to a clean branch.

**Review by kgaillot — COMMENTED** [review]
Nice, it's getting there

**Comment by kgaillot** [comment]
> Seeing the XML output reminds me that I forgot the schema portion of the crm_verify project :/ I'll open a new task for that ...

Actually, crm_verify XML output only has status and errors, so it fits in the existing schemas. So, all we need to do is update the crm_mon schema for this project once it's ready.

File: tools/crm_mon.c [hunk 5]
```diff
@@ -297,6 +297,7 @@ struct {
     { "summary", pcmk_section_summary },
     { "tickets", pcmk_section_tickets },
     { "times", pcmk_section_times },
+    { "verifications", pcmk_section_verify },
```
Comment (kgaillot) [inline tools/crm_mon.c:300]: This also needs to be added to two error messages -- search crm_mon.c for " options:"

**Review by kgaillot — COMMENTED** [review]
This looks great.

Add crm_mon regression tests -- you can use one of the crm_verify_invalid*xml files to test invalid syntax and crm_mon.xml to test valid syntax. Test text and XML for each using defaults, and also test --include=verifications for text+valid and --exclude=verifications for XML+invalid

Is there anything left you're unsure about?

**Comment by kgaillot** [comment]
> Add crm_mon regression tests -- you can use one of the crm_verify_invalid*xml files to test invalid syntax and crm_mon.xml to test valid syntax. Test text and XML for each using defaults, and also test --include=verifications for text+valid and --exclude=verifications for XML+invalid

Actually we have plenty of text/XML + valid tests, so we just need to add tests for invalid and --include/--exclude

File: lib/pacemaker/pcmk_output.c [hunk 6]
```diff
@@ -2358,41 +2340,58 @@ result_code_xml(pcmk__output_t *out, va_list args)
...
+    scheduler->priv = verify_out;
+
     verify_rc = pcmk__verify(scheduler, verify_out, scheduler->input);
+
+    pe_free_working_set(scheduler);
     pcmk__output_free(verify_out);
 
     if (verify_rc == pcmk_rc_ok) {
         if (pcmk_is_set(section_opts, pcmk_section_verify)) {
+            PCMK__OUTPUT_LIST_HEADER(out, false, rc, "Cluster Summary");
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:2365]: The way this macro works is that rc has to be initialized to pcmk_rc_no_output, and the macro will output the header the first time it's called and change rc to pcmk_rc_ok. Later calls will do nothing since the rc is changed. So, all calls for the same header have to use the same rc variable, which is why the header checks are all in the cluster-summary implementations.

This is a little different since we always want to output errors. Probably the easiest approach is to separate the verification itself from the output. We can always do the verification in the cluster-summary implementations, then output the message if the flag is set or there are errors.

**Review by kgaillot — COMMENTED** [review]
Getting there! When you're close to happy with it, flatten the code commits and the test output commits for the next review. Thanks!

File: lib/pacemaker/pcmk_output.c [hunk 7]
```diff
@@ -2337,12 +2337,72 @@ result_code_xml(pcmk__output_t *out, va_list args)
...
+
+    (void)(verify_rc);
+    (void)(section_opts);
+
+    pcmk__output_new(&verify_out, "none", NULL, NULL);
+
+    scheduler = pe_new_working_set();
+    scheduler->priv = verify_out;
+
+    verify_rc = pcmk__verify(scheduler, verify_out, scheduler->input);
```
[the remark this answers is not in the document]
Comment (waltdisgrace) [reply lib/pacemaker/pcmk_output.c:2358]: When I use the old scheduler object, there are some different failures:

* Failed (rc=000): crm_resource   - Try to move a resource to its existing location
* Failed (rc=064): crm_resource   - Move a resource from its existing location
* Failed (rc=108): crm_resource   - Move dummy to node1

File: lib/pacemaker/pcmk_output.c (hunk 6 above)
Comment (waltdisgrace) [reply lib/pacemaker/pcmk_output.c:2365]: To make sure I'm understanding this correctly, do you think the verification (`pcmk__verify()`) call should happen in cluster-summary and the output (`PCMK__OUTPUT_LIST_HEADER()` and `out->list_item()`) should happen in cluster-verify?

File: lib/pacemaker/pcmk_output.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_output.c:2358]: Calling `pcmk__verify()` shouldn't have any effects on the scheduler object, not sure how that's possible

[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_output.c:2358]: try rebasing on current main and see if that changes anything

File: lib/pacemaker/pcmk_output.c [hunk 8]
```diff
@@ -1999,6 +2002,21 @@ cluster_status_xml(pcmk__output_t *out, va_list args)
                      false);
     }
 
+
+    /* If there are verification errors, always print a statement about that, even if not requested */
+
+    pcmk__output_new(&verify_out, "none", NULL, NULL);
+    verify_rc = pcmk__verify(scheduler, verify_out, scheduler->input);
+    pcmk__output_free(verify_out);
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c]: You also need to set `scheduler->priv = verify_out` before calling `pcmk__verify()` (and save and restore the original value) because the scheduler code will output messages there too (`pcmk__output_cluster_status()` sets `scheduler->priv` before calling the cluster-status message)

File: lib/pacemaker/pcmk_output.c [hunk 9]
```diff
@@ -1999,6 +2002,21 @@ cluster_status_xml(pcmk__output_t *out, va_list args)
...
+
+    if (verify_rc == pcmk_rc_ok) {
+        if (pcmk_is_set(section_opts, pcmk_section_verify)) {
+            out->info(out, "CIB syntax is valid");
+        }
+    } else {
+        out->info(out, "CIB syntax has errors (for details, run crm_verify -LV).");
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c]: info() does nothing for XML, you have to add XML elements or attributes with the info

File: lib/pacemaker/pcmk_output.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_output.c:2358]: also ditto for `scheduler->priv`, that could be causing output that's changing the tests

File: lib/pacemaker/pcmk_output.c (hunk 6 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_output.c:2365]: I was thinking cluster-summary should call `pcmk__verify()` and output the header, and cluster-verify would just output the message based on the rc. however at that point we might as well just do everything in cluster-summary and not have a separate message for cluster-verify

File: lib/pengine/pe_output.c [hunk 10]
```diff
@@ -450,6 +458,28 @@ cluster_summary(pcmk__output_t *out, va_list args) {
...
+
+    if (verify_rc == pcmk_rc_ok) {
+        if (pcmk_is_set(section_opts, pcmk_section_verify)) {
+            PCMK__OUTPUT_LIST_HEADER(out, false, rc, "Cluster Summary");
+            out->list_item(out, NULL, "CIB syntax is valid");
+        }
+    } else {
+        /* If there are verification errors, always print a statement about that, even if not requested */
+        PCMK__OUTPUT_LIST_HEADER(out, false, rc, "Cluster Summary");
+        out->list_item(out, NULL, "CIB syntax has errors (for details, run crm_verify -LV)");
```
Comment (kgaillot) [inline lib/pengine/pe_output.c:480]: looking at this again, let's omit the crm_verify arguments -- the user could be running against a CIB_file. it's up to the user to figure out what arguments they need.

File: lib/pengine/pe_output.c [hunk 11]
```diff
@@ -450,6 +458,28 @@ cluster_summary(pcmk__output_t *out, va_list args) {
                      scheduler->localhost, last_written, user, client, origin);
     }
 
+    // Use the existing scheduler, but avoid scheduler output
+    pcmk__output_new(&verify_out, "none", NULL, NULL);
+    //scheduler = pe_new_working_set();
+    priv_orig = scheduler->priv;
+    scheduler->priv = verify_out;
+
+    verify_rc = pcmk__verify(scheduler, verify_out, scheduler->input);
```
Comment (kgaillot) [inline lib/pengine/pe_output.c:467]: What's happening is that `pcmk__verify()` calls `pcmk__schedule_actions()`, and then crm_simulate calls `pcmk__schedule_actions()` again, and it's not happy with being called twice.

One of these two approaches should work:

* Call `pe_reset_working_set()` after running `pcmk__verify()`. That will free `scheduler->input`, so you'll have to make a copy of that first and set it back afterward.
* Create a new scheduler object and copy the `input` and `now` members.

**Review by kgaillot — COMMENTED** [review]
I think I see what's happening ...

**Comment by kgaillot** [comment]
This should be revisited at some point (probably after T873, T882, and T727), but the code base has changed significantly enough to close this for now and open a new PR then.
